File: Example/OpAdxDemo_iOS/OpAdxDemo_iOS/Classes/LogView.m
//
//  LogView.m
//  OpAdxDemo_iOS
//
//  Created by Luan Chen on 2025/12/10.
//

#import "LogView.h"

@interface LogView ()

@property (nonatomic, strong, nullable) UITextView *textView;
@property (nonatomic, strong, nullable) UIScrollView *scrollView;

@end

@implementation LogView

- (void)setupInView:(UIView *)view {
    self.scrollView = [[UIScrollView alloc] init];
    self.textView = [[UITextView alloc] init];
    
    self.textView.editable = NO;
    self.textView.scrollEnabled = NO;
    self.textView.font = [UIFont systemFontOfSize:12];
    self.textView.backgroundColor = [UIColor systemGray6Color];
    
    [self.scrollView addSubview:self.textView];
    [view addSubview:self.scrollView];
}

- (void)setupConstraintsInView:(UIView *)view belowAnchorView:(UIView *)anchorView {
    if (!self.scrollView || !self.textView) return;
    
    self.scrollView.translatesAutoresizingMaskIntoConstraints = NO;
    self.textView.translatesAutoresizingMaskIntoConstraints = NO;
    
    UILayoutGuide *safeArea = view.safeAreaLayoutGuide;
    
    [NSLayoutConstraint activateConstraints:@[
        [self.scrollView.topAnchor constraintEqualToAnchor:anchorView.bottomAnchor constant:20],
        [self.scrollView.leadingAnchor constraintEqualToAnchor:view.leadingAnchor constant:20],
        [self.scrollView.trailingAnchor constraintEqualToAnchor:view.trailingAnchor constant:-20],
        [self.scrollView.bottomAnchor constraintEqualToAnchor:safeArea.bottomAnchor constant:-20],
        
        [self.textView.topAnchor constraintEqualToAnchor:self.scrollView.topAnchor],
        [self.textView.leadingAnchor constraintEqualToAnchor:self.scrollView.leadingAnchor],
        [self.textView.trailingAnchor constraintEqualToAnchor:self.scrollView.trailingAnchor],
        [self.textView.bottomAnchor constraintEqualToAnchor:self.scrollView.bottomAnchor],
        [self.textView.widthAnchor constraintEqualToAnchor:self.scrollView.widthAnchor]
    ]];
}

- (void)print:(NSString *)message {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (!self.textView || !self.scrollView) return;
        
        NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
        dateFormatter.dateFormat = @"HH:mm:ss";
        NSString *timestamp = [dateFormatter stringFromDate:[NSDate date]];
        
        NSString *logMessage = [NSString stringWithFormat:@"[%@] %@\n", timestamp, message];
        
        // 保存当前文本并追加新内容
        NSString *oldText = self.textView.text ?: @"";
        self.textView.text = [oldText stringByAppendingString:logMessage];
        
        // 强制布局更新
        [self.textView layoutIfNeeded];
        
        // 计算并设置滚动位置
        CGFloat textHeight = self.textView.contentSize.height;
        CGFloat scrollHeight = self.scrollView.bounds.size.height;
        
        if (textHeight > scrollHeight) {
            CGPoint bottomOffset = CGPointMake(0, textHeight - scrollHeight + 20);
            [self.scrollView setContentOffset:bottomOffset animated:NO];
        }
    });
}

@end

File: Example/OpAdxDemo_Swift/OpAdxSdkDemo/LogView.swift
//
//  LogView.swift
//  OperaAdxSDK
//
//  Created by Luan Chen on 2025/11/4.
//

// Supporting Types and Protocols
import UIKit

class LogView {
    private var textView: UITextView?
    private var scrollView: UIScrollView?
    
    func setup(in view: UIView) {
        let scrollView = UIScrollView()
        let textView = UITextView()
        
        textView.isEditable = false
        textView.isScrollEnabled = false
        textView.font = .systemFont(ofSize: 12)
        textView.backgroundColor = .systemGray6
        
        scrollView.addSubview(textView)
        view.addSubview(scrollView)
        
        self.textView = textView
        self.scrollView = scrollView
    }
    
    func setupConstraints(in view: UIView, below anchorView: UIView) {
        guard let scrollView = scrollView, let textView = textView else { return }
        
        scrollView.translatesAutoresizingMaskIntoConstraints = false
        textView.translatesAutoresizingMaskIntoConstraints = false
        
        NSLayoutConstraint.activate([
            scrollView.topAnchor.constraint(equalTo: anchorView.bottomAnchor, constant: 20),
            scrollView.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 20),
            scrollView.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -20),
            scrollView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -20),
            
            textView.topAnchor.constraint(equalTo: scrollView.topAnchor),
            textView.leadingAnchor.constraint(equalTo: scrollView.leadingAnchor),
            textView.trailingAnchor.constraint(equalTo: scrollView.trailingAnchor),
            textView.bottomAnchor.constraint(equalTo: scrollView.bottomAnchor),
            textView.widthAnchor.constraint(equalTo: scrollView.widthAnchor)
        ])
    }
    
    func print(_ message: String) {
        DispatchQueue.main.async {
            guard let textView = self.textView, let scrollView = self.scrollView else { return }
            
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "HH:mm:ss"
            let timestamp = dateFormatter.string(from: Date())
            
            let logMessage = "[\(timestamp)] \(message)\n"
            
            // 保存当前文本并追加新内容
            let oldText = textView.text ?? ""
            textView.text = oldText + logMessage
            
            // 强制布局更新
            textView.layoutIfNeeded()
            
            // 计算并设置滚动位置
            let textHeight = textView.contentSize.height
            let scrollHeight = scrollView.bounds.height
            
            if textHeight > scrollHeight {
                let bottomOffset = CGPoint(x: 0, y: textHeight - scrollHeight + 20)
                scrollView.setContentOffset(bottomOffset, animated: false)
            }
        }
    }
}
